### ISSUE TITLE
#524 follow-up

### ISSUE THREAD

**Issue by SirCmpwn** [swaywm/wlroots#527]
Opening a ticket so it doesn't get forgotten.

**Comment by acrisci** [comment]
In terms of the client api, this is what we want and the change accomplishes those goals: a single commit event for the surface where the role commit is guaranteed to have run first.

This approach mirrors Weston's, which is my usual starting point.

The alternative to this is to have two separate commit events for the surface: `surface::commit` and `surface::role_commit` where `role_commit` emits first.

**Comment by SirCmpwn** [comment]
I think my major beef with this is the roll-your-own callback logic rather than wl_signal. I can live with the rest of the changes.

**Comment by emersion** [comment]
The intent of the callback is to make sure there's only one listener (because a surface has only one role). This doesn't change much about type safety. I guess a wl_signal would be OK too.

**Comment by SirCmpwn** [comment]
Maybe it would be better to build out a `wlr_surface_role` abstraction that uses the same interface pattern we've done for i.e. wlr_output, wlr_pointer, etc.

**Comment by emersion** [comment]
An issue with this proposal is that role types (`wlr_output_cursor`, `wlr_wl_shell_surface`...) don't wrap a `wlr_surface`, so you can't get back the role type struct from the generic `wlr_surface`.

**Comment by SirCmpwn** [comment]
I don't understand how that follows.

**Comment by emersion** [comment]
Let's say you want to use a role interface and implement it for `wlr_wl_shell_surface`. The interface will look like:

```c
struct wlr_surface_role {
    const char *name;
    void (*commit)(struct wlr_surface *surface);
};
```

In `wlr_wl_shell.c` you'll write something like:

```c
static void wl_shell_surface_commit(struct wlr_surface *surface) {
    // Here I have a `wlr_surface`, but I want a `wlr_wl_shell_surface`
}

static const struct wlr_surface_role wl_shell_surface_role = {
    .name = "wl-shell-surface",
    .commit = wl_shell_surface_commit,
};
```

You'll need the `wlr_wl_shell_surface` in `wl_shell_surface_commit`.

**Comment by SirCmpwn** [comment]
We already encounter this in other places.

```c
struct wlr_wl_shell_surface {
    struct wlr_surface wlr_surface;
    // ...wl_shell state...
};

struct wlr_surface *wlr_surface = (struct wlr_surface*)wlr_wl_shell_surface;
struct wlr_wl_shell_surface *wlr_wl_shell_surface = (struct wlr_wl_shell_surface*)wlr_surface;
```

**Comment by emersion** [comment]
Yeah, so we can't do that because the surface is created before a role is assigned to it.

**Comment by SirCmpwn** [comment]
Sorry, rather:

```c
struct wlr_surface_role {
  struct wlr_surface *surface;
  // ...
};

struct wlr_wl_shell_surface_role {
  struct wlr_surface_role role;
  // ...wl_shell state...
};

struct wlr_surface {
   // ...
   struct wlr_surface_role *role;
   // ...
};
```

**Comment by emersion** [comment]
So the `wlr_surface_role` isn't `static const` anymore?

**Comment by SirCmpwn** [comment]
Was it ever?

**Comment by emersion** [comment]
All interfaces right now do this:

```c
static const struct wlr_surface_role wl_shell_surface_role = {
    .name = "wl-shell-surface",
    .commit = wl_shell_surface_commit,
};
```

But it seems that you're proposing to have one different `struct wlr_surface_role` per surface. That doesn't change a lot from @acrisci's PR, apart from putting stuff in a separate struct.

**Comment by SirCmpwn** [comment]
Yes, that's all it really does.

**Comment by SirCmpwn** [comment]
Well, in a manner consistent with our interfaces design.
